rfcs/20201221-tfmot-compression-api.md

```diff
@@ -0,0 +1,447 @@
+
+Compressed model contains the `decompress_weights` function in the graph. It’s possible to call the `decompress_weights` for each inference step. To improve performance, we’ll cache the decompressed one depending on flags if we have enough space.
+
+```python
+class WeightCompressionAlgorithm(metaclass=abc.ABCMeta):
+  """Interface for weight compression algorithm that acts on a per-layer basis.
+
+     This allows both options of either decompressing during inference or
+     decompressing prior to inference (where compression occurs by applying a
+     tool such as zip to the model file).
+
+     This interface is a purely functional one.
+  """
+
+  @abc.abstractmethod
+  def get_compressible_weights(
+      self, original_layer: tf.keras.layers.Layer) -> List[str]:
+    """Define compressible weights for each layer.
+
+    Args:
+       original_layer: tf.keras.layers.Layer representing a layer from the
+       original model.
+
+    Returns:
+       List of attribute names as string representing list of compressible
+       weights for the given layer. (e.g. return value ['kernel'] means
+       layer.kernel is compressible.)
+    """
+
+  @abc.abstractmethod
+  def init_training_weights(
+      self, pretrained_weight: tf.Tensor):
+    """Initialize training weights for the training model. It calls the `add_training_weight` method several times to add training weights.
+
+    Args:
+      pretrained_weight: tf.Tensor of a pretrained weight of a layer that will
+        be compressed eventually.
+    """
+
+  def add_training_weight(
+      self, *args, **kwargs):
+    """Add training weight for the training model. This method is called from `init_training_weights`.
+
+    Args:
+      *args, **kwargs: args and kwargs for training_model.add_weight.
+    """
+
+  @abc.abstractmethod
+  def project_training_weights(self, *training_weights: tf.Tensor) -> tf.Tensor:
+    """Define a piece of the forward pass during training, which operates on a single compressible weight.
+    The default throws an error when training occurs.
+
+    Args:
+       *training_weights: tf.Tensors representing any variables used during
+         training, for a single compressible weight, in the order returned in
+         `init_training_weights`.
+
+    Returns:
+       tf.Tensor to set the compressible weight to.
+    """
+
+  def update_training_weight(self, index: integer, tensor: tf.Tensor):
+    """Update a training weight on an index to a given tensor value.
+
+    This method is for the case that training weight should update to specific
+    value not from the model optimizer. It'll throws an error if it can't
+    find the training weight.
+
+    Args:
+      index: integer indicates index of training weight to update.
+      tensor: tf.Tensor to update specific training weight.
+    """
+
+  @abc.abstractmethod
+  def compress_training_weights(self, *training_weights: tf.Tensor) -> List[tf.Tensor]:
```

Comment:
Prefer having a single argument, which could be either a list of tensors or a single tensor in the case where you have a single tensor.

Reply:
I think it just require more lines for algorithms.

(e.g. Pruning case,

Current: 
def compress_training_weights(self, weight, mask):
  return weight * mask

The input argument changed to a list:
def compress_training_weights(self, training_weights):
  weight, mask = training_weights
  return weight * mask
)
(I'm not so sure it's the right way for this kinds of abstract method.)

Are you preferred the latter?

Reply:
Ok, sounds fine then. Thanks for the clarification.